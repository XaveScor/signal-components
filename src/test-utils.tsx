import React, { StrictMode } from "react";
import { createTestCtx, TestCtx } from "@reatom/testing";
import { render } from "@testing-library/react";
import { reatomContext } from "@reatom/npm-react";

export function customRender(element: React.ReactElement) {
  const ctx = createTestCtx();
  const rendered = render(element, {
    wrapper: ({ children }) => (
      <StrictMode>
        <reatomContext.Provider value={ctx}>{children}</reatomContext.Provider>
      </StrictMode>
    ),
  });

  return { rendered, ctx };
}
