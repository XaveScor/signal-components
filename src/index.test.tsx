import { describe, test, expect, vi } from "vitest";
import { screen } from "@testing-library/react";
import { declareComponent } from "./index";
import { atom } from "@reatom/core";
import { customRender } from "./test-utils";

describe("declare component", () => {
  describe("prop", () => {
    test("primitive", () => {
      const Component = declareComponent<{ x: number }>(({ x }) => {
        const y = atom((ctx) => ctx.spy(x) + 1);

        return (ctx) => {
          return <div>{ctx.spy(y)}</div>;
        };
      });

      customRender(<Component x={1} />);

      expect(screen.queryByText("2")).toBeInTheDocument();
      expect(screen.queryByText("1")).not.toBeInTheDocument();
    });

    test("atom", () => {
      const Component = declareComponent<{ x: number }>(({ x }) => {
        const y = atom((ctx) => ctx.spy(x) + 1);

        return (ctx) => {
          return <div>{ctx.spy(y)}</div>;
        };
      });

      customRender(<Component x={atom(1)} />);

      expect(screen.queryByText("2")).toBeInTheDocument();
      expect(screen.queryByText("1")).not.toBeInTheDocument();
    });
  });

  describe("change prop", () => {
    const Component = declareComponent<{ x: number }>(({ x }) => {
      const y = atom((ctx) => ctx.spy(x) + 1);

      return (ctx) => {
        return <div>{ctx.spy(y)}</div>;
      };
    });

    test("primitive => primitive", () => {
      const { rendered } = customRender(<Component x={1} />);

      expect(screen.queryByText("2")).toBeInTheDocument();

      rendered.rerender(<Component x={2} />);
      expect(screen.queryByText("3")).toBeInTheDocument();
    });

    test("primitive => atom", () => {
      const { rendered } = customRender(<Component x={1} />);

      expect(screen.queryByText("2")).toBeInTheDocument();

      rendered.rerender(<Component x={atom(2)} />);
      expect(screen.queryByText("3")).toBeInTheDocument();
    });

    test("atom => atom", () => {
      const { rendered } = customRender(<Component x={atom(1)} />);

      expect(screen.queryByText("2")).toBeInTheDocument();

      rendered.rerender(<Component x={atom(2)} />);
      expect(screen.queryByText("3")).toBeInTheDocument();
    });

    test("atom => primitive", () => {
      const { rendered } = customRender(<Component x={atom(1)} />);

      expect(screen.queryByText("2")).toBeInTheDocument();

      rendered.rerender(<Component x={2} />);
      expect(screen.queryByText("3")).toBeInTheDocument();
    });
  });

  test("change don't execute init phase", () => {
    const count = vi.fn();
    const Component = declareComponent<{ x: number }>(({ x }) => {
      count();
      return (ctx) => {
        return <div>{ctx.spy(x)}</div>;
      };
    });

    const { rendered } = customRender(<Component x={1} />);
    expect(count).toHaveBeenCalledTimes(2);
    rendered.rerender(<Component x={2} />);
    expect(count).toHaveBeenCalledTimes(2);
    rendered.rerender(<Component x={3} />);
    expect(count).toHaveBeenCalledTimes(2);
  });

  test("optional props", () => {
    const Component = declareComponent<{ x?: number }>(({ x }) => {
      const y = atom((ctx) =>
        ctx.spy(x) !== undefined ? "defined1" : "undefined2",
      );

      return (ctx) => {
        return <div>{ctx.spy(y)}</div>;
      };
    });

    const { rendered } = customRender(<Component />);

    expect(screen.queryByText("undefined2")).toBeInTheDocument();

    rendered.rerender(<Component x={2} />);
    expect(screen.queryByText("defined1")).toBeInTheDocument();
  });
});
