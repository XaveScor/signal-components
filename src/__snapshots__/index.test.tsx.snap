// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`declare component > return jsx > base test 1`] = `
<DocumentFragment>
  <div>
    1
  </div>
</DocumentFragment>
`;

exports[`declare component > return jsx > base test 2`] = `
<DocumentFragment>
  <div>
    2
  </div>
</DocumentFragment>
`;
